### PR TITLE
Make keypair optional

### PR DESCRIPTION
**Pull request by SoManyHs**

### CODE REVIEW

File: ecs-cli/modules/cli/cluster/cluster_app_test.go [hunk 1]
```diff
@@ -677,3 +675,16 @@ func TestClusterPSTaskGetInfoFail(t *testing.T) {
 	_, err = clusterPS(context, newMockReadWriter(), mockECS)
 	assert.Error(t, err, "Expected error in cluster ps")
 }
+
+func setupHappyPathMocks(mockECS *mock_ecs.MockECSClient, mockCloudformation *mock_cloudformation.MockCloudformationClient) {
```
Comment (PettitWesley) [inline ecs-cli/modules/cli/cluster/cluster_app_test.go]: Why is it Happy? :P

Comment (SoManyHs) [reply ecs-cli/modules/cli/cluster/cluster_app_test.go]: HAPPINESS IS MANDATORY

But really, it's because these are the basic mocks needed for a successful cluster up call. I should rename this to include ClusterUp in the name.

**Comment by SoManyHs** [comment]
Testing:
```
## previous cluster up and down commands:

~/go/src/github.com/aws/amazon-ecs-cli(make_keypair_optional)$ ecs-cli up --cluster test-with-keypair --keypair hhh-2 --capability-iam
INFO[0000] Created cluster                               cluster=test-with-keypair region=us-west-2
INFO[0000] Waiting for your cluster resources to be created...
INFO[0001] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0061] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0121] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"

~/go/src/github.com/aws/amazon-ecs-cli(make_keypair_optional)$ ecs-cli down --cluster test-with-keypair
Are you sure you want to delete your cluster? [y/N]
y
INFO[0002] Waiting for your cluster resources to be deleted...
INFO[0002] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0062] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0123] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0183] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0213] Deleted cluster                               cluster=test-with-keypair


## cluster up without keypair
~/go/src/github.com/aws/amazon-ecs-cli(make_keypair_optional)$ ecs-cli up --cluster test-without-keypair --capability-iam
WARN[0000] You will not be able to SSH into your EC2 instances without a key pair.
INFO[0000] Created cluster                               cluster=test-without-keypair region=us-west-2
INFO[0001] Waiting for your cluster resources to be created...
INFO[0001] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0066] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0131] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0197] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"

## testing compose commands
~/sandbox/test_fields$ ecs-cli compose up --cluster test-without-keypair
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0000] Using ECS task definition                     TaskDefinition="ecscompose-test_fields:20"
INFO[0000] Starting container...                         container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql"
INFO[0000] Starting container...                         container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress"
INFO[0000] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0000] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0012] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0012] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0024] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0024] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:20"
INFO[0030] Started container...                          container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:20"
INFO[0030] Started container...                          container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:20"

~/sandbox/test_fields$ ecs-cli compose down --cluster test-without-keypair
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0000] Stopping container...                         container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql"
INFO[0000] Stopping container...                         container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress"
INFO[0000] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=STOPPED lastStatus=RUNNING taskDefinition="ecscompose-test_fields:20"
INFO[0000] Describe ECS container status                 container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=STOPPED lastStatus=RUNNING taskDefinition="ecscompose-test_fields:20"
INFO[0006] Stopped container...                          container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/mysql" desiredStatus=STOPPED lastStatus=STOPPED taskDefinition="ecscompose-test_fields:20"
INFO[0006] Stopped container...                          container="2bc3fe05-4ebb-4a3a-b275-04a1f29344ba/wordpress" desiredStatus=STOPPED lastStatus=STOPPED taskDefinition="ecscompose-test_fields:20"

~/sandbox/test_fields$ ecs-cli compose service up --cluster test-without-keypair
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0000] Using ECS task definition                     TaskDefinition="ecscompose-test_fields:20"
INFO[0000] Created an ECS service                        service="ecscompose-service-test_fields" taskDefinition="ecscompose-test_fields:20"
INFO[0000] Updated ECS service successfully              desiredCount=1 serviceName="ecscompose-service-test_fields"
INFO[0015] Service status                                desiredCount=1 runningCount=1 serviceName="ecscompose-service-test_fields"
INFO[0015] (service ecscompose-service-test_fields) has started 1 tasks: (task afab257b-4df4-4735-b7cf-e76a42cb3fef).  timestamp=2017-10-24 17:49:27 +0000 UTC
INFO[0015] (service ecscompose-service-test_fields) has reached a steady state.  timestamp=2017-10-24 17:49:38 +0000 UTC
INFO[0015] ECS Service has reached a stable state        desiredCount=1 runningCount=1 serviceName="ecscompose-service-test_fields"

~/sandbox/test_fields$ ecs-cli compose service down --cluster test-without-keypair
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0005] Updated ECS service successfully              desiredCount=0 serviceName="ecscompose-service-test_fields"
INFO[0005] Service status                                desiredCount=0 runningCount=1 serviceName="ecscompose-service-test_fields"
INFO[0020] Service status                                desiredCount=0 runningCount=0 serviceName="ecscompose-service-test_fields"
INFO[0020] (service ecscompose-service-test_fields) has stopped 1 running tasks: (task afab257b-4df4-4735-b7cf-e76a42cb3fef).  timestamp=2017-10-24 17:50:40 +0000 UTC
INFO[0020] ECS Service has reached a stable state        desiredCount=0 runningCount=0 serviceName="ecscompose-service-test_fields"
INFO[0020] Deleted ECS service                           service="ecscompose-service-test_fields"
INFO[0020] ECS Service has reached a stable state        desiredCount=0 runningCount=0 serviceName="ecscompose-service-test_fields"

## cluster down without keypair
~/sandbox/test_fields$ ecs-cli down --cluster test-without-keypair
Are you sure you want to delete your cluster? [y/N]
y
INFO[0002] Waiting for your cluster resources to be deleted...
INFO[0002] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0062] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0123] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0183] Deleted cluster                               cluster=test-without-keypair
```

File: ecs-cli/modules/cli/cluster/cluster_app_test.go (hunk 1 above)
Comment (PettitWesley) [reply ecs-cli/modules/cli/cluster/cluster_app_test.go]: > HAPPINESS IS MANDATORY

ha ha.
> But really, it's because these are the basic mocks needed for a successful cluster up call. I should rename this to include ClusterUp in the name.

Cool!

File: ecs-cli/modules/clients/aws/cloudformation/params_test.go [hunk 2]
```diff
@@ -91,6 +85,7 @@ func TestAddWithUsePreviousValueAndValidate(t *testing.T) {
 	}
 
 	for _, param := range params {
+		fmt.Println(param)
```
Comment (PettitWesley) [inline ecs-cli/modules/clients/aws/cloudformation/params_test.go]: Why is this being printed? I'm guessing it was for debug?

File: ecs-cli/modules/clients/aws/cloudformation/template.go [hunk 3]
```diff
@@ -113,7 +113,7 @@ var template = `
       "ConstraintDescription": "must be a valid EC2 instance type."
     },
     "KeyName": {
-      "Type": "AWS::EC2::KeyPair::KeyName",
+      "Type": "String",
```
Comment (PettitWesley) [inline ecs-cli/modules/clients/aws/cloudformation/template.go]: Why the change in type?

File: ecs-cli/modules/clients/aws/cloudformation/params_test.go [hunk 4]
```diff
@@ -70,11 +64,11 @@ func TestAddAndValidate(t *testing.T) {
 
 	clusterValue, exists := cfnParams.nameToKeys[ParameterKeyCluster]
 	if !exists {
-		t.Errorf("Expcted key %s does not exist", ParameterKeyCluster)
+		t.Errorf("Expected key %s does not exist", ParameterKeyCluster)
```
Comment (PettitWesley) [inline ecs-cli/modules/clients/aws/cloudformation/params_test.go]: lol its kind of sad how many typos we have sitting around in the code; glad you're fixing them :)

File: ecs-cli/modules/clients/aws/cloudformation/template.go [hunk 5]
```diff
@@ -470,50 +461,16 @@ var template = `
...
-        "ImageId": { "Ref" : "EcsAmiId" },
-        "InstanceType": {
-          "Ref": "EcsInstanceType"
-        },
-        "AssociatePublicIpAddress": {
-          "Ref": "AssociatePublicIpAddress"
-        },
-        "IamInstanceProfile": {
-          "Ref": "EcsInstanceProfile"
-        },
```
Comment (PettitWesley) [inline ecs-cli/modules/clients/aws/cloudformation/template.go]: May be I'm reading the diff wrong, but why are all these lines being removed?

File: ecs-cli/modules/clients/aws/cloudformation/params_test.go (hunk 2 above)
Comment (SoManyHs) [reply ecs-cli/modules/clients/aws/cloudformation/params_test.go]: ....crap.

File: ecs-cli/modules/clients/aws/cloudformation/template.go (hunk 3 above)
Comment (SoManyHs) [reply ecs-cli/modules/clients/aws/cloudformation/template.go]: This is the same strategy the Console uses in their cloud formation template as well. The `AWS::EC2::KeyPair::KeyName` type apparently goes through special [validation](http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#AWS-Specific Parameter Types), which means the condition for checking if this value is an empty string won't work. Changing it to a string allows us to us the cloudformation conditions to set Keyname on the LaunchConfiguration

File: ecs-cli/modules/clients/aws/cloudformation/template.go (hunk 5 above)
Comment (SoManyHs) [reply ecs-cli/modules/clients/aws/cloudformation/template.go]: Previously, we had a condition to either create `EcsInstanceLc` if a keypair exists or a `EcsInstanceLcWithoutKeyPair` if it didn't. The latter resource was actually never created bc we were requiring keypair as a parameter.  This is now refactored to be similar to the cloudformation template used by the console team, which, rather than having a condition that decides which entire LaunchConfiguration resource gets created, just conditionally sets the KeyName field on one Launch Configuration template.

File: ecs-cli/modules/clients/aws/cloudformation/template.go (hunk 3 above)
Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/cloudformation/template.go]: Got it 👍

File: ecs-cli/modules/clients/aws/cloudformation/template.go (hunk 5 above)
Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/cloudformation/template.go]: Cool 👍

Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/cloudformation/template.go]: Oh, also, I don't know if you already tested this, but make sure that this change doesn't break backwards compatibility with old cloudformation templates and cause the down command to not work. Other than that PR looks good and I'll approve.

**Comment by SoManyHs** [comment]
Since the only change was for the LaunchConfiguration resource, and that was created based on a keypair, CloudFormation should still be able to locate that resource and delete it from the stack accordingly.

Just tested and cluster down appears to work with the new template:
```
~/go/src/github.com/aws/amazon-ecs-cli(dev)$ make
./scripts/build_binary.sh ./bin/local
Built ecs-cli

~/sandbox/test_fields$ ecs-cli up --cluster test-keypair-optional --capability-iam --keypair hhh-2
INFO[0000] Created cluster                               cluster=test-keypair-optional region=us-west-2
INFO[0001] Waiting for your cluster resources to be created... 
INFO[0001] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0061] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0121] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"
INFO[0182] Cloudformation stack status                   stackStatus="CREATE_IN_PROGRESS"

~/go/src/github.com/aws/amazon-ecs-cli(dev)$ git co -
Switched to branch 'make_keypair_optional'

~/go/src/github.com/aws/amazon-ecs-cli(make_keypair_optional)$ make
./scripts/build_binary.sh ./bin/local
Built ecs-cli

~/sandbox/test_fields$ ecs-cli down --cluster test-keypair-optional 
Are you sure you want to delete your cluster? [y/N]
y
INFO[0037] Waiting for your cluster resources to be deleted... 
INFO[0038] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0098] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0168] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0234] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0294] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0355] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0415] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0475] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0541] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0601] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0662] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0727] Cloudformation stack status                   stackStatus="DELETE_IN_PROGRESS"
INFO[0788] Deleted cluster                               cluster=test-keypair-optional
```
